allennlp/models/reading_comprehension/util.py

```diff
@@ -5,31 +5,29 @@ def get_best_span(span_start_logits: torch.Tensor, span_end_logits: torch.Tensor
     """
     This acts the same as the static method ``BidirectionalAttentionFlow.get_best_span()``
     in ``allennlp/models/reading_comprehension/bidaf.py``. We keep it here so that users can
-    directly import this function without the class. This function can be used to find the
-    best span with the highest probability from a passage, given the unnormalized scores
-    ``span_start_logits`` and ``span_end_logits``.
+    directly import this function without the class.
+
+    We call the inputs "logits" - they could either be unnormalized logits or normalized log
+    probabilities.  A log_softmax operation is a constant shifting of the entire logit
+    vector, so taking an argmax over either one gives the same result.
     """
     if span_start_logits.dim() != 2 or span_end_logits.dim() != 2:
         raise ValueError("Input shapes must be (batch_size, passage_length)")
     batch_size, passage_length = span_start_logits.size()
-    max_span_log_prob = [-1e20] * batch_size
-    span_start_argmax = [0] * batch_size
-    best_word_span = span_start_logits.new_zeros((batch_size, 2), dtype=torch.long)
-
-    span_start_logits = span_start_logits.detach().cpu().numpy()
-    span_end_logits = span_end_logits.detach().cpu().numpy()
-
-    for b in range(batch_size):  # pylint: disable=invalid-name
-        for j in range(passage_length):
-            val1 = span_start_logits[b, span_start_argmax[b]]
-            if val1 < span_start_logits[b, j]:
-                span_start_argmax[b] = j
-                val1 = span_start_logits[b, j]
-
-            val2 = span_end_logits[b, j]
+    device = span_start_logits.device
+    # (batch_size, passage_length, passage_length)
+    span_log_probs = span_start_logits.unsqueeze(2) + span_end_logits.unsqueeze(1)
+    # Only the upper triangle of the span matrix is valid; the lower triangle has entries where
+    # the span ends before it starts.
+    span_log_mask = torch.triu(torch.ones((passage_length, passage_length),
+                                          device=device)).log().unsqueeze(0)
```

Comment:
If I do that, the `span_log_mask` would be two-dimensional, which does not match with the `span_log_probs`?

Reply:
There was a comment on the previous PR saying it wasn't necessary.  I just removed it, and we'll see if it crashes...  If it does, I'll add it back.

Reply:
Pytorch might be smart enough to do the broadcasting; I'm not sure in this case if it is or not.  I suppose it could possibly lead to some crazy edge case, maybe...  I was just trying to avoid an additional operation.

Reply:
Looks like tests pass without it.

Reply:
Ok, you already did that. That's fine.